src/crafting.cpp

```diff
@@ -42,268 +43,134 @@ recipe::recipe() :
+
+        if( sk.empty() ) {
+            // clear all requirements
+
+        } else if( sk.has_array( 0 ) ) {
+            // multiple requirements
+            while( sk.has_more() ) {
+                auto arr = sk.next_array();
+                r.required_skills[skill_id( arr.get_string( 0 ) )] = arr.get_int( 1 );
+            }
+
+        } else {
+            // single requirement
+            r.required_skills[skill_id( sk.get_string( 0 ) )] = sk.get_int( 1 );
         }
     }
 
-    auto req_id = std::string( "inline_recipe_" ) += rec_name;
-    requirement_data::load_requirement( jsobj, req_id );
-    rec->reqs.emplace_back( requirement_id( req_id ), 1 );
+    // simplified autolearn sets requirements equal to required skills at finalization
+    if( jo.has_bool( "autolearn" ) ) {
+        assign( jo, "autolearn", r.autolearn );
 
-    jsarr = jsobj.get_array( "book_learn" );
-    while( jsarr.has_more() ) {
-        JsonArray ja = jsarr.next_array();
-        recipe::bookdata_t bd{ ja.get_string( 0 ), ja.get_int( 1 ), "", false };
-        if( ja.size() >= 3 ) {
-            bd.recipe_name = ja.get_string( 2 );
-            bd.hidden = bd.recipe_name.empty();
+    } else if( jo.has_array( "autolearn" ) ) {
+        r.autolearn = false;
+        auto sk = jo.get_array( "autolearn" );
+        while( sk.has_more() ) {
+            auto arr = sk.next_array();
+            r.autolearn_requirements[skill_id( arr.get_string( 0 ) )] = arr.get_int( 1 );
         }
-        rec->booksets.push_back( bd );
     }
 
-    // Note, a recipe has to be fully instantiated before adding
-    recipe_dict.add( rec );
-}
+    if( jo.has_member( "decomp_learn" ) ) {
+        r.learn_by_disassembly.clear();
 
-void reset_recipes()
-{
-    recipe_dict.clear();
-}
-
-void finalize_recipes()
-{
-    // Has to be done before recipe finalization!
-    for( auto r : recipe_dict ) {
-        r->requirements_ = std::accumulate( r->reqs.begin(), r->reqs.end(), requirement_data(),
-        []( const requirement_data & lhs, const std::pair<requirement_id, int> &rhs ) {
-            return lhs + ( *rhs.first * rhs.second );
-        } );
-    }
-
-    recipe_dict.finalize();
-
-    std::ostringstream buffer;
-    for( auto r : recipe_dict ) {
-        buffer.clear();
-        for( auto j = r->booksets.begin(); j != r->booksets.end(); ++j ) {
-            const std::string &book_id = j->book_id;
-            if( !item::type_is_defined( book_id ) ) {
-                buffer << "book " << book_id << " for recipe " << r->ident() << " does not exist" << "\n";
-                continue;
-            }
-            const itype *t = item::find_type( book_id );
-            if( !t->book ) {
-                // TODO: we could make up a book slot?
-                buffer << "book " << book_id << " for recipe " << r->ident() << " is not a book" << "\n";
-                continue;
+        if( jo.has_int( "decomp_learn" ) ) {
+            if( !r.skill_used ) {
```

Comment:
What is the `!` operator? The usage is ambiguous here - does it mean that the skill is set or is valid?
Check for validity should be in verification/finalization, to avoid needless order dependence.